src/ServiceControl/Recoverability/Retries/Infrastructure/ReturnToSenderDequeuer.cs

```diff
@@ -1,25 +1,120 @@
         {
             this.sender = sender;
+
+            faultManager = new CaptureIfMessageSendingFails(store, bus, ExecuteOnFailure);
         }
 
         protected override void HandleMessage(TransportMessage message)
         {
-            var destinationAddress = Address.Parse(message.Headers["ServiceControl.TargetEndpointAddress"]);
+            var destination = message.Headers["ServiceControl.TargetEndpointAddress"];
+            var destinationAddress = Address.Parse(destination);
 
             message.Headers.Remove("ServiceControl.TargetEndpointAddress");
             message.Headers.Remove("ServiceControl.Retry.StagingId");
 
-            sender.Send(message, destinationAddress);
+            try
+            {
+                sender.Send(message, destinationAddress);
+            }
+            catch (Exception)
+            {
+                message.Headers["ServiceControl.TargetEndpointAddress"] = destination;
+
+                throw;
+            }
+        }
+
+        protected override IManageMessageFailures FaultManager
+        {
+            get { return faultManager; }
+        }
+
+        class CaptureIfMessageSendingFails : IManageMessageFailures
+        {
+            static ILog Log = LogManager.GetLogger(typeof(CaptureIfMessageSendingFails));
+            private IDocumentStore store;
+            private IBus bus;
+            readonly Action executeOnFailure;
+
+            public CaptureIfMessageSendingFails(IDocumentStore store, IBus bus, Action executeOnFailure)
+            {
+                this.store = store;
+                this.bus = bus;
+                this.executeOnFailure = executeOnFailure;
+            }
+
+            public void SerializationFailedForMessage(TransportMessage message, Exception e)
+            {
+            }
+
+            public void ProcessingAlwaysFailsForMessage(TransportMessage message, Exception e)
+            {
+                try
+                {
+                    var destination = message.Headers["ServiceControl.TargetEndpointAddress"];
+                    var messageUniqueId = message.Headers["ServiceControl.Retry.UniqueMessageId"];
+                    Log.Error(string.Format("Failed to send '{0}' message to '{1}' for retry.", messageUniqueId, destination), e);
+
+                    var key = FailedMessage.MakeDocumentId(messageUniqueId);
+                    store.DatabaseCommands.Patch(key,
+                        new[]
+                        {
+                            new PatchRequest
+                            {
+                                Type = PatchCommandType.Set,
+                                Name = "Status",
+                                Value = (int) FailedMessageStatus.Unresolved,
+                            }
+                        });
+
+                    store.DatabaseCommands.Delete(FailedMessageRetry.MakeDocumentId(messageUniqueId), null);
```

Comment:
Do these two database commands respect transaction scope? Ideally we want both or neither of them to work.